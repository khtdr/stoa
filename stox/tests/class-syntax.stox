#!/usr/bin/env stox

class Breakfast {
  cook(food) {
    print "Fried eggs and " + food + ".";
  }
  serve(who) {
    print "Enjoy, " + who + ".";
  }
}

class Bagel {}
var bagel = Bagel();

bagel.isReady = false;
print bagel.isReady ? "Bagel is ready" : "... still toasting...";
bagel.isReady = true;
print bagel.isReady ? "Bagel is ready" : "... still toasting...";

class Standard {}

var object = Standard();
object.name = "Standard";
object.meal = Breakfast();

var newBagel = Bagel();
object.meal.toast = newBagel;

object.meal.toast.isReady = false;
print newBagel.isReady ? "Bagel is ready" : "... still toasting...";
newBagel.isReady = true;
print object.meal.toast.isReady ? "Bagel is ready" : "... still toasting...";

object.meal.cook(object.meal.toast);
object.meal.serve('customer');

class Bacon { eat() { print "Crunch crunch crunch"; } }
Bacon().eat();
