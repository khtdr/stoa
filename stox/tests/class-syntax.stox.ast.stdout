(program 
   (class Breakfast
      (fun cook (let [food] (block 
         (print (+ (+ "Fried eggs and " food) "."))
      )))
      (fun serve (let [who] (block 
         (print (+ (+ "Enjoy, " who) "."))
      )))
   )
   (class Bagel
      
   )
   (var bagel (call Bagel))
   (.set bagel isReady false)
   (print (?: (.get bagel isReady) "Bagel is ready" "... still toasting..."))
   (.set bagel isReady true)
   (print (?: (.get bagel isReady) "Bagel is ready" "... still toasting..."))
   (class Standard
      
   )
   (var object (call Standard))
   (.set object name "Standard")
   (.set object meal (call Breakfast))
   (var newBagel (call Bagel))
   (.set (.get object meal) toast newBagel)
   (.set (.get (.get object meal) toast) isReady false)
   (print (?: (.get newBagel isReady) "Bagel is ready" "... still toasting..."))
   (.set newBagel isReady true)
   (print (?: (.get (.get (.get object meal) toast) isReady) "Bagel is ready" "... still toasting..."))
   (call (.get (.get object meal) cook) (.get (.get object meal) toast))
   (call (.get (.get object meal) serve) "customer")
   (class Bacon
      (fun eat (let [] (block 
         (print "Crunch crunch crunch")
      )))
   )
   (call (.get (call Bacon) eat))
)
